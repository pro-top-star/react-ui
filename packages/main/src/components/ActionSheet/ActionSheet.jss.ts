const styles = {
  /*
   *  is being applied to the encapsulating ul element
   */
  actionSheet: {
    listStyleType: 'none',
    margin: 0,
    padding: '0.1875rem 0.375rem',
    '&$tablet,&$phone': {
      padding: '0.25rem 0.5rem'
    },
    '& ui5-button': {
      display: 'block'
    }
  },
  phone: {},
  tablet: {}
};

export default styles;
