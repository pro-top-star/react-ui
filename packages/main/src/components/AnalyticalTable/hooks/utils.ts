export const makeTemplateColumns = (columns, resizedColumns) => {
  return columns
    .map((column) => {
      if (resizedColumns[column.id]) {
        return `${resizedColumns[column.id]}px`;
      }
      return `minmax(${column.minWidth}px, ${column.width || '1fr'})`;
    })
    .join(' ');
};
