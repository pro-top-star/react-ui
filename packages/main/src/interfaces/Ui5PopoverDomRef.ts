import { Ui5DomRef } from './Ui5DomRef';

export interface Ui5PopoverDomRef extends Ui5DomRef {
  openBy: (p: HTMLElement) => void;
  close: () => void;
}
