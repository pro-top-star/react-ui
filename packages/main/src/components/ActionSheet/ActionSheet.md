```jsx
import { ActionSheet } from '@ui5/webcomponents-react/lib/ActionSheet';
```
