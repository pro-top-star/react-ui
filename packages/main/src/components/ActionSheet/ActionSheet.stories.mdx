import { Meta, Title, Description, Story, Canvas, ArgsTable } from '@storybook/addon-docs/blocks';
import { createSelectArgTypes } from '@shared/stories/createSelectArgTypes';
import { ActionSheet } from '@ui5/webcomponents-react/lib/ActionSheet';
import { Button } from '@ui5/webcomponents-react/lib/Button';
import { PlacementType } from '@ui5/webcomponents-react/lib/PlacementType';
import { PopoverHorizontalAlign } from '@ui5/webcomponents-react/lib/PopoverHorizontalAlign';
import { PopoverVerticalAlign } from '@ui5/webcomponents-react/lib/PopoverVerticalAlign';
import { useCallback, useRef } from 'react';
import { DocsHeader } from '@shared/stories/DocsHeader';

<Meta
  title="Components / ActionSheet"
  component={ActionSheet}
  argTypes={{
    ...createSelectArgTypes({
      placementType: PlacementType,
      horizontalAlign: PopoverHorizontalAlign,
      verticalAlign: PopoverVerticalAlign
    }),
    children: {
      type: null
    },
    footer: {
      type: null
    },
    header: {
      type: null
    },
    ref: {
      type: null
    }
  }}
  args={{
    horizontalAlign: PopoverHorizontalAlign.Center,
    placementType: PlacementType.Right,
    verticalAlign: PopoverVerticalAlign.Center
  }}
/>

<DocsHeader />

<Canvas>
  <Story name="Default">
    {(args) => {
      const actionSheetRef = useRef();
      const onButtonClick = useCallback(
        (e) => {
          actionSheetRef.current.open(e.target);
        },
        [actionSheetRef]
      );
      return (
        <>
          <Button onClick={onButtonClick}>Open ActionSheet</Button>
          <ActionSheet ref={actionSheetRef} {...args}>
            <Button icon="add">Accept</Button>
            <Button>Reject</Button>
            <Button>This is my super long text!</Button>
          </ActionSheet>
        </>
      );
    }}
  </Story>
</Canvas>

<ArgsTable story="." />
