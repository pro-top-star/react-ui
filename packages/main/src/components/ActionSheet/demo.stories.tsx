import { select } from '@storybook/addon-knobs';
import { ActionSheet } from '@ui5/webcomponents-react/lib/ActionSheet';
import { Button } from '@ui5/webcomponents-react/lib/Button';
import { PlacementType } from '@ui5/webcomponents-react/lib/PlacementType';
import React from 'react';

export default {
  title: 'Components / ActionSheet',
  component: ActionSheet
};

export const defaultStory = () => (
  <ActionSheet
    openBy={<Button>Open ActionSheet</Button>}
    placement={select('placement', [PlacementType.Top, PlacementType.Bottom], PlacementType.Top)}
  >
    <Button icon="add">Accept</Button>
    <Button>Reject</Button>
    <Button>This is my super long text!</Button>
  </ActionSheet>
);

defaultStory.story = { name: 'Default' };
