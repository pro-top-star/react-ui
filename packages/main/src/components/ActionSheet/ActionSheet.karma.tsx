import { mountThemedComponent } from '@shared/tests/utils';
import { expect, use } from 'chai';
import { matchSnapshot } from 'chai-karma-snapshot';
import React, { createRef, RefObject } from 'react';
import { Ui5PopoverDomRef } from '../../interfaces/Ui5PopoverDomRef';
import { ActionSheet } from '../../lib/ActionSheet';
import { Button } from '../../lib/Button';

use(matchSnapshot);

describe('ActionSheet', () => {
  it('Render without Crashing', () => {
    const button = <Button />;
    const wrapper = mountThemedComponent(<ActionSheet openBy={button} />);
    expect(wrapper.debug()).to.matchSnapshot();
  });

  it('Test Legacy Ref', () => {
    let legacyRef = null;

    const ref = (el) => {
      legacyRef = el;
    };
    const button = <Button />;
    mountThemedComponent(<ActionSheet ref={ref} openBy={button} />);
    expect(legacyRef.tagName).to.equal('UI5-POPOVER');
  });

  it('Ref object', () => {
    const ref: RefObject<Ui5PopoverDomRef> = createRef();
    const button = <Button />;
    mountThemedComponent(<ActionSheet ref={ref} openBy={button} />);
    expect((ref.current as any).tagName).to.equal('UI5-POPOVER');
  });

  // it('Click Handler', () => {
  //   // @ts-ignore
  //   const callback = sinon.spy();
  //   const wrapper = mountThemedComponent(
  //     // @ts-ignore
  //     <ActionSheet openBy={<Button />}>
  //       // @ts-ignore
  //       <Button onClick={callback}>Test</Button>
  //     </ActionSheet>
  //   );
  //   // (wrapper.find('Button').last().prop('onPress') as any)({target: {}});
  //   wrapper.find('div').first().simulate('click');
  //   wrapper.update();
  //   wrapper.find('Button').last().simulate('click');
  //   // (wrapper.find('Button').last().prop('onPress') as any)({target: {}});
  //   expect(callback.called).to.equal(true);
  // });
  //
  // it('Close by outside click', () => {
  //   const callback = sinon.spy();
  //   const wrapper = mountThemedComponent(
  //     // @ts-ignore
  //     <ActionSheet openBy={<Button>Test</Button>}>
  //       // @ts-ignore
  //       <Button onClick={callback}>Test</Button>
  //     </ActionSheet>
  //   );
  //   wrapper
  //     .find('Button')
  //     .first()
  //     .simulate('click');
  //   wrapper.update();
  //   // @ts-ignore
  //   const component = wrapper.find('ActionSheet').instance();
  //   // @ts-ignore
  //   expect(component.state.open).to.equal(true);
  //
  //   // @ts-ignore
  //   component.onPopoverClosed();
  //   wrapper.update();
  //   // @ts-ignore
  //   expect(component.state.open).to.equal(false);
  // });
});
