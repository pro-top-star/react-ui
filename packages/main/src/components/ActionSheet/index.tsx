import { addCustomCSS } from '@ui5/webcomponents-base/dist/Theming';
import { Device } from '@ui5/webcomponents-react-base/lib/Device';
import { StyleClassHelper } from '@ui5/webcomponents-react-base/lib/StyleClassHelper';
import { useConsolidatedRef } from '@ui5/webcomponents-react-base/lib/useConsolidatedRef';
import { usePassThroughHtmlProps } from '@ui5/webcomponents-react-base/lib/usePassThroughHtmlProps';
import { ButtonDesign } from '@ui5/webcomponents-react/lib/ButtonDesign';
import { PlacementType } from '@ui5/webcomponents-react/lib/PlacementType';
import { Popover } from '@ui5/webcomponents-react/lib/Popover';
import React, { Children, cloneElement, FC, forwardRef, ReactElement, ReactNode, RefObject } from 'react';
import { createUseStyles } from 'react-jss';
import { CommonProps } from '../../interfaces/CommonProps';
import { Ui5PopoverDomRef } from '../../interfaces/Ui5PopoverDomRef';
import { ButtonPropTypes } from '../../webComponents/Button';
import styles from './ActionSheet.jss';

export interface ActionSheetPropTypes extends CommonProps {
  openBy: ReactNode;
  placement?: PlacementType;
  children?: ReactElement<ButtonPropTypes> | ReactElement<ButtonPropTypes>[];
}

const useStyles = createUseStyles(styles, { name: 'ActionSheet' });

addCustomCSS(
  'ui5-button',
  `
  :host([data-is-action-sheet-button]) .ui5-button-root {
    justify-content: flex-start;
  }
  `
);

/**
 * <code>import { ActionSheet } from '@ui5/webcomponents-react/lib/ActionSheet';</code>
 */
const ActionSheet: FC<ActionSheetPropTypes> = forwardRef(
  (props: ActionSheetPropTypes, ref: RefObject<Ui5PopoverDomRef>) => {
    const { children, placement, openBy, style, slot, className } = props;

    const classes = useStyles();

    const actionSheetClasses = StyleClassHelper.of(classes.actionSheet);
    if (className) {
      actionSheetClasses.put(className);
    }
    if (Device.system.tablet) {
      actionSheetClasses.put(classes.tablet);
    } else if (Device.system.phone) {
      actionSheetClasses.put(classes.phone);
    }

    const popoverRef: RefObject<Ui5PopoverDomRef> = useConsolidatedRef(ref);

    const onActionButtonClicked = (handler) => (e) => {
      popoverRef.current.close();
      if (typeof handler === 'function') {
        handler(e);
      }
    };

    const renderActionSheetButton = (element, index) => {
      return cloneElement(element, {
        key: index,
        design: ButtonDesign.Transparent,
        onClick: onActionButtonClicked(element.props.onClick),
        'data-is-action-sheet-button': ''
      });
    };

    const passThroughProps = usePassThroughHtmlProps(props);

    return (
      <Popover
        ref={popoverRef}
        openBy={openBy}
        placementType={placement}
        style={style}
        slot={slot}
        className={actionSheetClasses.valueOf()}
        {...passThroughProps}
      >
        {Children.map(children, renderActionSheetButton)}
      </Popover>
    );
  }
);

ActionSheet.defaultProps = {
  placement: PlacementType.Bottom
};
ActionSheet.displayName = 'ActionSheet';

export { ActionSheet };
