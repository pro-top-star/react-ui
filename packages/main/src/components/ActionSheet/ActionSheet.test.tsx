import { createPassThroughPropsTest, mountThemedComponent } from '@shared/tests/utils';
import { ActionSheet } from '@ui5/webcomponents-react/lib/ActionSheet';
import { Label } from '@ui5/webcomponents-react/lib/Label';
import { Button } from '@ui5/webcomponents-react/lib/Button';
import React, { createRef, RefObject } from 'react';
import { Ui5PopoverDomRef } from '../../interfaces/Ui5PopoverDomRef';
import sinon from 'sinon';

describe('ActionSheet', () => {
  test('Render without Crashing', () => {
    const button = <Button />;
    const wrapper = mountThemedComponent(
      <ActionSheet openBy={button} className="myCustomClass">
        <Button icon={'add'}>Accept</Button>
        <Button>Reject</Button>
        <Button>This is my super long text!</Button>
      </ActionSheet>
    );
    expect(wrapper.render()).toMatchSnapshot();
  });

  test('Button Click handler', () => {
    const callback = sinon.spy();
    const button = <Button />;
    const wrapper = mountThemedComponent(
      <ActionSheet openBy={button} className="myCustomClass">
        <Button icon={'add'} onClick={callback}>
          Accept
        </Button>
        <Button>Reject</Button>
        <Button>This is my super long text!</Button>
      </ActionSheet>
    );
    wrapper
      .find('ui5-button')
      .first()
      .instance()
      // @ts-ignore
      .fireEvent('click');
    wrapper.update();

    wrapper
      .find('ui5-popover ui5-button')
      .first()
      .instance()
      // @ts-ignore
      .fireEvent('click');
    expect(callback.called).toBe(true);
  });

  test('Test Legacy Ref', () => {
    let legacyRef = null;

    const ref = (el) => {
      legacyRef = el;
    };
    const button = <Button />;
    mountThemedComponent(
      <ActionSheet ref={ref} openBy={button}>
        <Button icon={'add'}>Accept</Button>
        <Button>Reject</Button>
        <Button>This is my super long text!</Button>
      </ActionSheet>
    );
    expect(legacyRef.tagName).toEqual('UI5-POPOVER');
  });

  test('Ref object', () => {
    const ref: RefObject<Ui5PopoverDomRef> = createRef();
    const button = <Button />;
    mountThemedComponent(<ActionSheet ref={ref} openBy={button} />);
    expect((ref.current as any).tagName).toEqual('UI5-POPOVER');
  });

  test('does not crash with other component', () => {
    const button = <Button />;
    const wrapper = mountThemedComponent(
      <ActionSheet openBy={button}>
        <Label>I should not crash</Label>
      </ActionSheet>
    );
    expect(wrapper.render()).toMatchSnapshot();
  });

  createPassThroughPropsTest(ActionSheet);
});
