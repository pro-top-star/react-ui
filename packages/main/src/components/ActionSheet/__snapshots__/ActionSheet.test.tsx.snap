// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionSheet Render without Crashing 1`] = `
Array [
  <div
    style="display: inline-block;"
  >
    <ui5-button
      design="Default"
    />
  </div>,
  <ui5-popover
    horizontal-align="Center"
    placement-type="Bottom"
    vertical-align="Center"
  >
    <ul
      class="ActionSheet--actionSheet-"
    />
  </ui5-popover>,
]
`;
